File: iteration-2/src/inputCreator.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.regex.Pattern;

public class inputCreator {


    public inputCreator() {

        JSONParser namesP = new JSONParser();
        JSONParser courseP = new JSONParser();
        ArrayList<Object> AllCourses = new ArrayList<>();

        try {

            JSONObject inputC = (JSONObject) courseP.parse(new FileReader("iteration-2/src/courses.json"));
            JSONObject inputN = (JSONObject) namesP.parse(new FileReader("iteration-2/src/names.json"));
            int quota = 10;
            int counter = 0;

            //System.out.println(AllCourses);
            // parse the JSON data and convert it to a JSONObject
            //get all names from names.json
            JSONObject output = new JSONObject(); // get input file using output name. this jsonobject uses names.json and courses.json and create input.json.


            // get the array of names from the JSONObject
            JSONArray names = (JSONArray) inputN.get("names");
            JSONArray inputCourses = (JSONArray) inputC.get("courses");

            System.out.println(AllCourses);
            // loop through the array of names

            File file = new File("input.json");
            file.delete();

            FileWriter writer1 = new FileWriter("input.json", true);
            writer1.write("{\n" + '"' +"AllStudents" + '"' + ":[\n");
            writer1.flush();

            for (Object name : names) {
                JSONArray passedcoursesarray = new JSONArray();
                JSONArray failedcoursearray = new JSONArray();
                JSONArray PassedCoursesPrereq = new JSONArray();
                if (output.get("Name") != name) {
                    int totalgivenCredits = 0;
                    output.put("Name", name);
                    // give student name. because every human being has name.

                    // Then add the "CurrentSemester" and "PassedCoursesCode" elements
                    output.put("CurrentSemester", 1 + (int) +(Math.random() * 8)); // give student random currentsemester.
                    int currentSemester = 0;
                    for (Object c : inputCourses) {
                        double randomNumber = Math.random(); // for chance

                        int result = 0;
                        currentSemester = (int) output.get("CurrentSemester");
                        double chance = currentSemester * 0.01;

                        //öğrencinin ders geçme şansını dönemine göre %10 arttır.
                        if (randomNumber < 0.5 - chance) {
                            result = 0;
                        } else {
                            result = 1;
                        }
                        int grade = 0;
                        String lettergrade;
                        JSONObject course = (JSONObject) c;
                        String courseCode = (String) course.get("CourseCode");
                        String courseName = (String) course.get("CourseName");
                        String preRequisite = course.get("PrerequisiteCourseCodes").toString().replaceAll("[^a-zA-Z0-9]", "");
                        int courseSemester = Integer.parseInt(course.get("courseSemester").toString());
                        int credits = Integer.parseInt(course.get("Credit").toString());

                        if ((int) output.get("CurrentSemester") > courseSemester) {
                            if (result == 1) {
                                totalgivenCredits += credits;
                                if (!preRequisite.equals("")) {
                                    PassedCoursesPrereq.add(preRequisite);
                                    for (int i = 0; i < passedcoursesarray.size(); i++) {
                                        if (preRequisite.equals(passedcoursesarray.get(i))) {
                                            passedcoursesarray.add(courseCode);
                                            grade = (int) ((Math.random() * (100 - 50)) + 50);
                                            passedcoursesarray.add("Grade: " + grade);
                                        }
                                    }
                                } else {
                                    passedcoursesarray.add(courseCode);
                                    grade = (int) ((Math.random() * (100 - 50)) + 50);
                                    passedcoursesarray.add("Grade: " + grade);
                                    switch (grade / 10) {
                                        case 10:
                                            lettergrade = "A+";
                                            break;
                                        case 9:
                                            lettergrade = "A";
                                            break;
                                        case 8:
                                            lettergrade = "A-";
                                            break;
                                        case 7:
                                            lettergrade = "B";
                                            break;
                                        case 6:
                                            lettergrade = "C";
                                            break;
                                        case 5:
                                            lettergrade = "D";
                                            break;
                                        default:
                                            lettergrade = "Invalid grade!";
                                            break;
                                    }
                                    passedcoursesarray.add("LetterGrade: " + lettergrade);
                                }
                            } else {
                                failedcoursearray.add(courseCode);
                                grade = (int) ((Math.random() * (50)) + 0);
                                failedcoursearray.add(grade);
                                switch (grade / 10) {
                                    case 4:
                                        lettergrade = "FF";
                                        break;
                                    case 3:
                                        lettergrade = "FF";
                                        break;
                                    case 2:
                                        lettergrade = "FF";
                                        break;
                                    case 1:
                                        lettergrade = "FF";
                                        break;
                                    default:
                                        lettergrade = "FF";
                                        break;
                                }
                                failedcoursearray.add("LetterGrade: " + lettergrade);
                            }
                        }
                        output.put("PassedCoursesCode", passedcoursesarray);
                        output.put("FailedCoursesCode", failedcoursearray);
                        output.put("preRequisite", PassedCoursesPrereq);
                        output.put("TotalGivenCredits", totalgivenCredits);
                    }


                    int randomNumber2 = (int) (Math.random() * 2); // for entryYear
                    int randomNumber3 = 100 + (int) (Math.random() * 900); // for student id
                    double v = currentSemester / 2.0;
                    int entryYear = (int) (2022 - Math.ceil(v) - randomNumber2);
                    String studentID = "1501" + (int) (entryYear - 2000) + (randomNumber3);

                    String studentmail = output.get("Name").toString().toLowerCase().replace(" ", "") + entryYear + "@marun.edu.tr";
                    output.put("entryYear", (int) entryYear);
                    output.put("studentID", studentID);
                    output.put("studentMail", studentmail);
                    JSONObject jsonObject = new JSONObject();
                    System.out.println(output); // just checkin.
                    try (FileWriter writer = new FileWriter("input.json", true)) {
                        counter++;
                        if (counter != 999) {
                            writer.write(output.toJSONString() + ",\n");
                        } else {
                            writer.write(output.toJSONString() + "\n");

                        }


                        writer.flush();
                    } catch (IOException e) {
                        System.out.println("Error writing to file: " + e.getMessage());
                    }

                    // Reset the output object to an empty object
                    output = new JSONObject();
                }
            }
            FileWriter writer2 = new FileWriter("input.json", true);
            writer2.write("]}");

            writer2.flush();
            System.out.printf("Created %d student.", counter);
        } catch (IOException e) {
            System.out.println("Error reading JSON file: " + e.getMessage());
        } catch (ParseException e) {
            System.out.println("Error parsing JSON file: " + e.getMessage());
        }
    };

}

File: iteration-3/InputCreator.py
import json
import random

def inputCreator():
    # Read in the names and courses JSON files
    with open("names.json", encoding="utf-8") as f:
       names = json.load(f,)
    with open("courses.json", encoding="utf-8") as f:
       courses = json.load(f)
       

    all_students = []


    # Loop through the names
    for name in names["names"]:
        passed_courses = []
        failed_courses = []
        passed_courses_prereq = []
        total_given_credits = 0

        Enteryear = random.choice([2017,2018,2019,2020,2021,2022])
        year = 2022 - Enteryear 
        inc = random.choice([0,1])
        current_semester= (year*2)+inc
        
        if current_semester > 8 :
            current_semester = 8
        
        random_number_3 = 100 + random.randint(0, 900)
        student_id = "1501" + str(Enteryear- 2000) + str(random_number_3)
        

        
        
        for course in courses["courses"]:
            if course["courseSemester"]>current_semester:
                continue
            # Generate a random number between 0 and 1
            chance = random.uniform(0, 1)
            
            
            if chance < 0.5 - (current_semester * 0.01):
                # Student has failed the course
                failed_courses.append({
                    "code": course["CourseCode"],
                    "name": course["CourseName"],
                    "grade" : int(random.uniform(0, 50)),
                    "letterGrade": random.choice(["FF"])
                })
            else:
                # Student has passed the course
                total_given_credits += course["Credit"]
                grade = int(random.uniform(50, 100))
                lettergrade = "AA", "AB", "BB" , "BC" ,"CC","DC","DD"
                if grade >= 90:
                    lettergrade = "AA"
                elif grade >= 85:
                    lettergrade = "AB"
                elif grade >= 80:
                    lettergrade = "BA"
                elif grade >= 75:
                    lettergrade = "CB"
                elif grade >= 65:
                    lettergrade = "CC"
                elif grade >= 55:
                    lettergrade = "DC"
                elif grade >= 50:
                    lettergrade = "DD"
                passed_courses.append({
                    "code": course["CourseCode"],
                    "name": course["CourseName"],
                    "grade": grade,
                    "letterGrade": lettergrade
                })
                if course["PrerequisiteCourseCodes"]:
                    passed_courses_prereq.extend(course["PrerequisiteCourseCodes"])
                for i in range(len(passed_courses)):
                    if course["PrerequisiteCourseCodes"] == passed_courses[i]["code"]:
                        passed_courses_prereq.remove(course["PrerequisiteCourseCodes"])
                        break

        # Boş stringleri sil
        preRequisite = [val for val in passed_courses_prereq if val != ""]             
             
        # Create the student object
        student = {
            "Name": name,
            "StudentId":student_id,
            "preRequisite" : preRequisite,
            "Enteryear": Enteryear,
            "CurrentSemester": current_semester,
            "PassedCourses": passed_courses,
            "FailedCourses": failed_courses,
            "TotalGivenCredits": total_given_credits
        }
        
        # Add the student object to the list
        all_students.append(student)

    # Create the final output JSON object
    output = {
        "AllStudents": all_students
    }

    # Write the output to a file
    with open("input.json", "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False)


inputCreator()


